src/Analysis/Engine/Impl/Interpreter/Ast/AstAnalysisFunctionWalker.cs

```diff
@@ -52,7 +51,7 @@ AstPythonFunctionOverload overload
             var classmethodObj = _scope.Interpreter.GetBuiltinType(BuiltinTypeId.ClassMethod);
             var staticmethodObj = _scope.Interpreter.GetBuiltinType(BuiltinTypeId.StaticMethod);
             foreach (var d in (_target.Decorators?.Decorators).MaybeEnumerate().ExcludeDefault()) {
-                var m = _scope.GetValueFromExpression(d);
+                var m = _scope.GetValueFromExpression(d).ResolveType();
```

Comment:
So what is a criteria of calling `ResolveType`?

Reply:
Generally in AST walk types should be lazy to the very end unless it is trivial to resolve. Otherwise, ResolveType should be called

- in delayed actions which are actually invoked to resolve type
- when code needs actual type, as the snippet above
- After AST walk I am forcibly resolving types in all scopes (pending change)